File: leetcode/494.target-sum.py
# @before-stub-for-debug-begin
from python3problem494 import *
from typing import *

# @before-stub-for-debug-end

#
# @lc app=leetcode id=494 lang=python3
#
# [494] Target Sum
#

# @lc code=start
class Solution:
    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        dp = {}  # (index,total) -> # of ways

        def dfs(i, total):
            if i == len(nums):
                return 1 if total == target else 0
            if (i, total) in dp:
                return dp[(i, total)]

            dp[(i, total)] = dfs(i + 1, total + nums[i]) + dfs(i + 1, total - nums[i])
            return dp[(i, total)]

        return dfs(0, 0)


# @lc code=end

File: leetcode/494.target-sum.java
import java.util.HashMap;
import java.util.Map;

/*
 * @lc app=leetcode id=494 lang=java
 *
 * [494] Target Sum
 */

// @lc code=start
class Solution {
    Map<String, Integer> dp = new HashMap<>();
    int[] nums;
    int target;

    public int findTargetSumWays(int[] nums, int target) {
        this.nums = nums;
        this.target = target;
        int ans = dfs(0, 0);
        // System.out.println(dp);
        return ans;
    }

    public int dfs(int i, int total) {
        if (i == nums.length) {
            return (total == target) ? 1 : 0;
        }
        String key = i + "()" + total;

        if (dp.containsKey(key)) {
            return dp.get(key);
        }

        dp.put(key, dfs(i + 1, total + nums[i]) + dfs(i + 1, total - nums[i]));
        return dp.get(key);
    }
}
// @lc code=end
